src/backtrace.rs

```diff
@@ -0,0 +1,113 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+use core::mem::size_of;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub enum BTReturnAddress {
+    ReturnAddress(u64),
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+        }
+    }
+}
+
+pub struct BackTrace {
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+        let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+
+    fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
+        let offset = (fp - self.stack_start - 2) as usize / size_of::<u64>();
```

Comment:
Needs to guard against potential underflows:

`let offset = fp.checked_sub(self.stack_start).unwrap_or(2) .checked_sub(2).unwrap_or(0) as usize / sizeof::u64();`

Reply:
It uses slice.get(), which returns None on out of bounds.

Reply:
Think we are referring to different portions of code unless the code needs to be refreshed. In the below, it seems offset is computed unconditionally, which can cause an underflow if `fp == self.stack_start` for example. The first checked_sub() guards against an underflow with `(fp - self.stack_start)`, and the second guards against the case where `'fp - self.stack_start'` evaluates to 0 (which will cause a underflow with the -2).

In either of those cases, offset will be 0.

   ```suggestion
         fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
             let offset = (fp.checked_sub(self.stack_start).unwrap_or(2).checked_sub(2).unwrap_or(0) as usize / size_of::<u64>();
```